File: leetcode/src/02256_minimum_average_difference/solution.rs
struct Solution;

impl Solution {
    pub fn minimum_average_difference(nums: Vec<i32>) -> i32 {
        let psum = nums
            .iter()
            .scan(0i64, |acc, &num| {
                *acc += num as i64;
                Some(*acc)
            })
            .collect::<Vec<_>>();
        (0..psum.len())
            .map(|i| (Self::difference(&psum, i), i))
            .min()
            .unwrap()
            .1 as i32
    }

    fn difference(psum: &[i64], idx: usize) -> i64 {
        if idx == psum.len() - 1 {
            return psum[psum.len() - 1] / psum.len() as i64;
        }
        (psum[idx] / (idx + 1) as i64
            - (psum[psum.len() - 1] - psum[idx]) / (psum.len() - 1 - idx) as i64)
            .abs()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let nums = vec![2, 5, 3, 9, 5, 3];
        assert_eq!(Solution::minimum_average_difference(nums), 3);

        let nums = vec![0];
        assert_eq!(Solution::minimum_average_difference(nums), 0);
    }
}

File: leetcode/src/02256_minimum_average_difference/solution.go
package leetcode

import "math"

func minimumAverageDifference(nums []int) int {
	psum := prefixSum(nums)
	curMinAverage := int64(math.MaxInt64)
	idx := -1
	for i := range psum {
		diff := difference(psum, i)
		if idx == -1 || diff < curMinAverage {
			curMinAverage = diff
			idx = i
		}
	}
	return idx
}

func prefixSum(nums []int) []int64 {
	res := make([]int64, len(nums))
	res[0] = int64(nums[0])
	for i := 1; i < len(nums); i++ {
		res[i] += int64(nums[i]) + res[i-1]
	}
	return res
}

func difference(psum []int64, idx int) int64 {
	if idx == len(psum)-1 {
		return psum[len(psum)-1] / int64(len(psum))
	}
	return int64(math.Abs(float64(psum[idx]/int64(idx+1) - (psum[len(psum)-1]-psum[idx])/int64(len(psum)-1-idx))))
}
